<a href="https://www.kaggle.com/code/avtnshm/wine-quality-xgboost-vs-catboost?scriptVersionId=157535110" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/red-wine-quality-cortez-et-al-2009/winequality-red.csv


In [2]:
df = pd.read_csv('/kaggle/input/red-wine-quality-cortez-et-al-2009/winequality-red.csv')
df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


# CatBoost

In [4]:
from catboost import CatBoostClassifier, CatBoostRegressor

In [5]:
X = df.drop("quality", axis=1)
y = df["quality"]

In [6]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [7]:
model = CatBoostRegressor()
model.fit(X_train, y_train)

Learning rate set to 0.042566
0:	learn: 0.7961722	total: 55.5ms	remaining: 55.5s
1:	learn: 0.7858269	total: 57.1ms	remaining: 28.5s
2:	learn: 0.7768657	total: 58.6ms	remaining: 19.5s
3:	learn: 0.7683315	total: 60.1ms	remaining: 15s
4:	learn: 0.7595828	total: 61.6ms	remaining: 12.3s
5:	learn: 0.7519151	total: 63.2ms	remaining: 10.5s
6:	learn: 0.7438187	total: 64.6ms	remaining: 9.17s
7:	learn: 0.7359962	total: 66ms	remaining: 8.18s
8:	learn: 0.7295845	total: 67.4ms	remaining: 7.42s
9:	learn: 0.7242132	total: 68.8ms	remaining: 6.81s
10:	learn: 0.7183437	total: 70.3ms	remaining: 6.32s
11:	learn: 0.7121187	total: 71.6ms	remaining: 5.9s
12:	learn: 0.7060758	total: 73.2ms	remaining: 5.56s
13:	learn: 0.7003074	total: 74.6ms	remaining: 5.25s
14:	learn: 0.6943954	total: 76.2ms	remaining: 5s
15:	learn: 0.6895442	total: 77.6ms	remaining: 4.77s
16:	learn: 0.6854317	total: 79ms	remaining: 4.57s
17:	learn: 0.6812536	total: 80.4ms	remaining: 4.38s
18:	learn: 0.6763736	total: 81.7ms	remaining: 4.22s
19

In [8]:
from sklearn.metrics import mean_squared_error
y_pred = model.predict(X_test)

rmse = np.sqrt(mean_squared_error(y_test, y_pred))
print("RMSE on test set:", rmse)

RMSE on test set: 0.5520607100823542


# XGBoost

In [9]:
import xgboost as xgb

In [10]:
xgb_model = xgb.XGBRegressor()

In [11]:
xgb_model.fit(X_train, y_train)

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=None, device=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             gamma=None, grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=None, max_bin=None,
             max_cat_threshold=None, max_cat_to_onehot=None,
             max_delta_step=None, max_depth=None, max_leaves=None,
             min_child_weight=None, missing=nan, monotone_constraints=None,
             multi_strategy=None, n_estimators=None, n_jobs=None,
             num_parallel_tree=None, random_state=None, ...)

In [12]:
y_pred = xgb_model.predict(X_test)

In [13]:
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
print(f"XGBoost RMSE on test set: {rmse}")

XGBoost RMSE on test set: 0.5926756553831612


# Hyperparameter Tuning

In [14]:
from sklearn.model_selection import GridSearchCV
param_grid = {
    'learning_rate': [0.01, 0.05, 0.1],
    'depth': [4, 6, 8],
    'iterations': [100, 200, 300],
}

grid_search = GridSearchCV(estimator=CatBoostRegressor(), param_grid=param_grid, cv=5)
grid_search.fit(X_train, y_train)

best_model = grid_search.best_estimator_
print(grid_search.best_params_)


0:	learn: 0.8078757	total: 960us	remaining: 95ms
1:	learn: 0.8055940	total: 1.69ms	remaining: 82.8ms
2:	learn: 0.8036895	total: 2.33ms	remaining: 75.5ms
3:	learn: 0.8017396	total: 3.02ms	remaining: 72.4ms
4:	learn: 0.7995745	total: 3.69ms	remaining: 70.2ms
5:	learn: 0.7973272	total: 4.35ms	remaining: 68.1ms
6:	learn: 0.7953070	total: 5.03ms	remaining: 66.8ms
7:	learn: 0.7931808	total: 5.62ms	remaining: 64.7ms
8:	learn: 0.7907990	total: 6.29ms	remaining: 63.6ms
9:	learn: 0.7886913	total: 6.96ms	remaining: 62.6ms
10:	learn: 0.7866829	total: 7.59ms	remaining: 61.5ms
11:	learn: 0.7845927	total: 8.32ms	remaining: 61ms
12:	learn: 0.7824840	total: 9.05ms	remaining: 60.5ms
13:	learn: 0.7805785	total: 9.76ms	remaining: 59.9ms
14:	learn: 0.7782816	total: 10.5ms	remaining: 59.5ms
15:	learn: 0.7762676	total: 11.2ms	remaining: 58.6ms
16:	learn: 0.7743706	total: 11.9ms	remaining: 58ms
17:	learn: 0.7725270	total: 12.5ms	remaining: 56.8ms
18:	learn: 0.7706946	total: 13.7ms	remaining: 58.4ms
19:	learn:

In [15]:
tuned_catboost_model = CatBoostRegressor(depth=8, iterations=300, learning_rate=0.1)

In [16]:
tuned_catboost_model.fit(X_train, y_train)

0:	learn: 0.7820833	total: 4.27ms	remaining: 1.28s
1:	learn: 0.7612164	total: 8.15ms	remaining: 1.21s
2:	learn: 0.7395001	total: 11.8ms	remaining: 1.17s
3:	learn: 0.7206845	total: 15ms	remaining: 1.11s
4:	learn: 0.7031643	total: 18.2ms	remaining: 1.07s
5:	learn: 0.6882901	total: 21.6ms	remaining: 1.06s
6:	learn: 0.6741201	total: 25ms	remaining: 1.04s
7:	learn: 0.6630111	total: 28.3ms	remaining: 1.03s
8:	learn: 0.6507977	total: 31.7ms	remaining: 1.02s
9:	learn: 0.6404128	total: 34.9ms	remaining: 1.01s
10:	learn: 0.6290752	total: 38ms	remaining: 1000ms
11:	learn: 0.6218547	total: 42.6ms	remaining: 1.02s
12:	learn: 0.6143574	total: 46.4ms	remaining: 1.02s
13:	learn: 0.6059994	total: 49.7ms	remaining: 1.01s
14:	learn: 0.5998445	total: 52.8ms	remaining: 1s
15:	learn: 0.5948182	total: 56ms	remaining: 994ms
16:	learn: 0.5881100	total: 59.6ms	remaining: 992ms
17:	learn: 0.5829676	total: 62.9ms	remaining: 985ms
18:	learn: 0.5766310	total: 66.3ms	remaining: 980ms
19:	learn: 0.5728266	total: 69.5

In [17]:
y_pred = tuned_catboost_model.predict(X_test)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
print("Tuned CatBoost RMSE:", rmse)

Tuned CatBoost RMSE: 0.5419016977440334


In [19]:
param_grid = {
    'eta': [0.1, 0.05, 0.01],
    'max_depth': [4, 6, 8],
    'subsample': [0.8, 1.0],
    'colsample_bytree': [0.8, 1.0],
    'n_estimators': [100, 200, 300],
    'gamma': [0, 1, 5]
}

In [20]:
dtrain = xgb.DMatrix(X_train, y_train)
grid_search = GridSearchCV(xgb.XGBRegressor(), param_grid, cv=5)

In [22]:
grid_search.fit(X_train, y_train) 
best_model = grid_search.best_estimator_
best_params = grid_search.best_params_
y_pred = best_model.predict(X_test)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
print(f"Tuned XGBoost RMSE:", rmse)

Tuned XGBoost RMSE: 0.5580942557523639


# Comparison b/w Tuned Models

### RMSE Score
* CatBoost: 0.542
* XGBoost: 0.558
* Observation: CatBoost has a slightly better RMSE, meaning its average prediction error is lower.